src/engine/GroupByImpl.h

```diff
@@ -591,9 +591,49 @@ class GroupByImpl : public Operation {
 
   // TODO<joka921> Also inform the query planner (via the cost estimate)
   // that the optimization can be done.
-};
 
-// _____________________________________________________________________________
+  FRIEND_TEST(GroupBySamplingTest, belowThreshold);
+  FRIEND_TEST(GroupBySamplingTest, aboveThreshold);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseEmptyInput);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseAllUnique);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseAllSame);
+
+ private:
+  // Check via sampling if the hash-map path should be skipped due to too many
+  // groups
+  bool shouldSkipHashMapGrouping(const IdTable& table) const;
+  bool shouldSkipHashMapGroupingDynamic(
+      const std::shared_ptr<const Result>& subresult) const;
```

Comment:
Have you already looked (etc. at the Service implementation, which also has to store parts of a lazy result), 
on how you can do this estimation for lazy results? This would be the most important feature here, as the subresults are often lazy, and it is only the sort that makes them non-lazy.